Improve the performance of detecting linux-musl platform by parsing ELF

ntkme opened the pull request:
Closes #314

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Read chunk of data from a file descriptor into a new Buffer.
+ */
+const readFileDescriptor = function (
```
Nit: for consistency:

```suggestion
function readFileDescriptor(
```

 I'll add a lint to enforce this.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+
+import * as fs from 'fs';
+
+/**
+ * Read chunk of data from a file descriptor into a new Buffer.
+ */
```
Nit:

```suggestion
/** Read a chunk of data from a file descriptor into a new Buffer. */
```

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+  return buffer;
+};
+
+/**
+ * Parse an ELF file and return its interpreter.
+ */
```
Nit: 
```suggestion
/** Parse an ELF file and return its interpreter. */
```

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Parse an ELF file and return its interpreter.
+ */
+export const getELFInterpreter = function (path: string): string {
```
Nit:

```suggestion
export function getElfInterpreter(path: string): string {
```

Even in JS code, we try to follow the Dart style of capitalizing 3+-letter acronyms as normal words.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+      eIdent.getUint8(2) !== 0x4c ||
+      eIdent.getUint8(3) !== 0x46
+    ) {
+      throw new Error(`${path} is not an elf file.`);
```
```suggestion
      throw new Error(`${path} is not an ELF file.`);
```

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+
+    const eIdentClass = eIdent.getUint8(4);
+    if (eIdentClass !== 1 && eIdentClass !== 2) {
+      throw new Error(`${path} has an invalid elf class.`);
```
```suggestion
      throw new Error(`${path} has an invalid ELF class.`);
```

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+
+    const e_phoff = class32
+      ? eIdent.getUint32(28, littleEndian)
+      : Number(eIdent.getBigUint64(32, littleEndian));
```
Probably worth adding a comment explaining why we believe it's safe to convert a bigint to a normal number here.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+    }
+    const littleEndian = eIdentData === 1;
+
+    const e_phoff = class32
```
Nit: Don't use underscores in variable names. Calling this `ePhoff` would be better, but since the fact that it's being used as an offset is relevant to the code, something explicit like `eProgHeaderOffset` would be even better.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+      ? eIdent.getUint16(44, littleEndian)
+      : eIdent.getUint16(56, littleEndian);
+
+    const proghdrs = new DataView(
```
Similarly, calling this something less abbreviated like `progHeaders` would be clearer.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+export const getELFInterpreter = function (path: string): string {
+  const fd = fs.openSync(path, 'r');
+  try {
+    const eIdent = new DataView(readFileDescriptor(fd, 0, 64).buffer);
```
What does this `e` indicate here?

ntkme replied:
In original kernel code this is named as `e_ident` (I think e stands for ELF):
 https://github.com/torvalds/linux/blob/21b136cc63d2a9ddd60d4699552b69c214b32964/include/uapi/linux/elf.h#L210C17-L210C24

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+    for (let i = 0; i < e_phnum; i++) {
+      const byteOffset = i * e_phentsize;
+      const p_type = proghdrs.getUint32(byteOffset, littleEndian);
+      if (p_type !== 3) {
+        continue;
+      }
```
Nit: make `if` statements single-line if possible.

```suggestion
      if (p_type !== 3) continue;
```

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+    for (let i = 0; i < e_phnum; i++) {
+      const byteOffset = i * e_phentsize;
+      const p_type = proghdrs.getUint32(byteOffset, littleEndian);
+      if (p_type !== 3) {
```
Add a brief comment explaining what "type 3" means.

nex3 commented on lib/src/elf.ts:
```diff
@@ -0,0 +1,99 @@
+      const p_offset = class32
+        ? proghdrs.getUint32(byteOffset + 4, littleEndian)
+        : Number(proghdrs.getBigUint64(byteOffset + 8, littleEndian));
+      const p_filesz = class32
```
What does the "z" indicate here?

ntkme replied:
This is the original variable name in linux kernel code: https://github.com/torvalds/linux/blob/21b136cc63d2a9ddd60d4699552b69c214b32964/include/uapi/linux/elf.h#L254

nex3 commented on lib/src/compiler-path.ts:
```diff
@@ -4,20 +4,22 @@
-const isLinuxMusl = function () {
-  return fs.readFileSync(process.execPath).includes('/ld-musl-');
+const isLinuxMusl = function (path: string): boolean {
+  const interpreter = getELFInterpreter(path);
```
I wonder if we should wrap this in a `try`/`catch`, print any error it produces to `stderr`, and just fall back to returning false if it errors. If the user is running musl, running the executable will error anyway, but if they're not that'll allow them to continue working without being strictly blocked on debugging why their ELF format isn't what we expect.

ntkme commented:
I mostly reused the variable names from the linux kernel code: https://github.com/torvalds/linux/blob/HEAD/include/uapi/linux/elf.h. Idea here is for whoever going to work on this code they can refer to the kernel API document and find the reference without too much effort: 

@nex3 Now I have renamed a few of them, and changed the snake case to camel case, let me know if you have any preferences on these naming.

nex3 commented:
I think it's better to have explicit names than names that exactly match the kernel, especially since the explicit names are just expanded versions of the names from the kernel anyway so figuring out that they correspond won't be too hard for a reader. I'd also like to avoid single-letter affixes with unclear meaning like `e`, `p`, and `z`.

nex3 reviewed (approved):
Thanks!

ntkme commented:
@nex3 Okay, I've renamed everything to something that direct represent what they mean.
